File: packages/shared/src/decrees.ts
/**
 * Royal Decrees, from War Chest: Nobility.
 *
 * Three of the seven are dealt face up at setup. A player proclaims by
 * discarding the Royal Coin face up, placing one of their three Proclamation
 * Seals on a Decree and carrying out its ability — so each side gets each
 * Decree once per game. A Decree may not be chosen if its ability cannot be
 * fully carried out.
 */

import { DECREE_IDS } from './generated.js';
import type { LocalizedText } from './units.js';

export type DecreeId =
  | 'sacrifice'
  | 'guard'
  | 'march'
  | 'enlist'
  | 'redeploy'
  | 'spy'
  | 'reinforce';

export interface DecreeDefinition {
  readonly id: DecreeId;
  readonly name: LocalizedText;
  readonly text: LocalizedText;
}

export const DECREES: Readonly<Record<DecreeId, DecreeDefinition>> = {
  sacrifice: {
    id: 'sacrifice',
    name: { en: 'Sacrifice', ru: 'Жертва' },
    text: {
      en: 'Attack with a friendly unit, then remove a coin from the friendly unit.',
      ru: 'Атакуйте дружественным отрядом, затем снимите с него монету.',
    },
  },
  guard: {
    id: 'guard',
    name: { en: 'Guard', ru: 'Стража' },
    text: {
      en: 'Attack with a friendly unit that is on a location you control.',
      ru: 'Атакуйте дружественным отрядом, который стоит на вашей локации.',
    },
  },
  march: {
    id: 'march',
    name: { en: 'March', ru: 'Марш' },
    text: {
      en: 'Move a friendly bolstered unit.',
      ru: 'Сделайте ход усиленным дружественным отрядом.',
    },
  },
  enlist: {
    id: 'enlist',
    name: { en: 'Enlist', ru: 'Вербовка' },
    text: { en: 'Recruit twice.', ru: 'Наймите две монеты.' },
  },
  redeploy: {
    id: 'redeploy',
    name: { en: 'Redeploy', ru: 'Передислокация' },
    text: {
      en: 'Remove a friendly unit from the board and then deploy it.',
      ru: 'Снимите дружественный отряд с поля и разверните его заново.',
    },
  },
  spy: {
    id: 'spy',
    name: { en: 'Spy', ru: 'Шпион' },
    text: {
      en: "Look at an opponent's hand. You may discard one of their coins. If you do, they draw a new coin.",
      ru: 'Посмотрите руку соперника. Можете сбросить одну его монету — тогда он берёт новую.',
    },
  },
  reinforce: {
    id: 'reinforce',
    name: { en: 'Reinforce', ru: 'Подкрепление' },
    text: {
      en: 'Take one of your coins that has been removed from play and place it in your supply.',
      ru: 'Верните в свой запас одну свою монету, выбывшую из игры.',
    },
  },
};

export { DECREE_IDS };

/** Decrees dealt face up for a game, and Seals each side may spend. */
export const DECREES_IN_PLAY = 3;
export const SEALS_PER_SIDE = 3;

/** A Decree on the table, with the sides that have already used it. */
export interface DecreeInPlay {
  readonly id: DecreeId;
  /** Teams whose Seal is already on this Decree. */
  seals: number[];
}

File: crates/wc-core/src/decrees.rs
//! Royal Decrees, from War Chest: Nobility.
//!
//! Three of the seven are dealt face up at setup. A player proclaims by
//! discarding the Royal Coin face up, placing one of their three Proclamation
//! Seals on a Decree and carrying out its ability — so each side gets each
//! Decree once per game.

#[derive(Copy, Clone, PartialEq, Eq, Debug, Hash)]
#[repr(u8)]
pub enum DecreeId {
    Sacrifice = 0,
    Guard,
    March,
    Enlist,
    Redeploy,
    Spy,
    Reinforce,
}

pub const DECREE_COUNT: usize = 7;

pub const DECREE_IDS: [DecreeId; DECREE_COUNT] = {
    use DecreeId::*;
    [Sacrifice, Guard, March, Enlist, Redeploy, Spy, Reinforce]
};

pub const DECREE_KEYS: [&str; DECREE_COUNT] =
    ["sacrifice", "guard", "march", "enlist", "redeploy", "spy", "reinforce"];

impl DecreeId {
    pub fn key(self) -> &'static str {
        DECREE_KEYS[self as usize]
    }

    pub fn from_key(key: &str) -> Option<DecreeId> {
        DECREE_KEYS.iter().position(|k| *k == key).map(|i| DECREE_IDS[i])
    }

    pub fn from_idx(i: usize) -> Option<DecreeId> {
        DECREE_IDS.get(i).copied()
    }
}

/// Decrees dealt face up for a game, and Seals each side may spend.
pub const DECREES_IN_PLAY: usize = 3;
pub const SEALS_PER_SIDE: u8 = 3;

/// A Decree on the table, with the sides that have already used it.
///
/// The teams are a list rather than a bitmask because the order is the order
/// they proclaimed in, and that order is part of the state a conformance run
/// compares byte for byte.
#[derive(Clone, PartialEq, Eq, Debug)]
pub struct DecreeInPlay {
    pub id: DecreeId,
    pub seals: arrayvec::ArrayVec<u8, 4>,
}

impl DecreeInPlay {
    pub fn new(id: DecreeId) -> DecreeInPlay {
        DecreeInPlay { id, seals: arrayvec::ArrayVec::new() }
    }

    #[inline]
    pub fn has_seal(&self, team: u8) -> bool {
        self.seals.contains(&team)
    }

    #[inline]
    pub fn add_seal(&mut self, team: u8) {
        self.seals.push(team);
    }
}
